src/c/uCLib_base

```diff
@@ -0,0 +1,127 @@
+/*
+ * RISC OS uCLib (Micro C Library)
+ * 
+ * purpouse: This is a small (micro) re-implementation of a generic
+ *           C Library to use on RISC OS to build utils (transient)
+ *           executables  using  the C  Programming  language  and,
+ *           hopefully, C++ as well.
+ *           Please note: This library is NOT a replacement for the
+ *           Acorn/ROOL  SharedCLib (SCL)  or  the  ANSILib (ANSL).
+ *           It is  intended for use  with applications that do not
+ *           need the  entire  SCL or ANSL,  that  wish to  have no
+ *           dependecies  on  the  SCL/ANSL and still use C for the
+ *           source coding.
+ *
+ *   author: Paolo Fabio Zaino
+ *  release: 0.0.0
+ * 
+ * Copyright 2021 by Paolo Fabio Zaino
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * Redistributions  of  source  code  must retain the above copyright
+ * notice,  this list of conditions, the distribution license and the
+ * following disclaimer.
+ * Redistributions   in   binary   form  must   reproduce  the  above
+ * copyright  notice,  this  list  of  conditions  and  the following
+ * disclaimer  in  the  documentation and/or other materials provided
+ * with the distribution.
+ * Neither  the  name of the author nor the names of the contributors
+ * may  be  used  to  endorse  or  promote products derived from this
+ * software without specific prior written permission.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License
+ * the restrictions above and the disclaimer below.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * THIS  SOFTWARE   IS  PROVIDED  BY   THE  COPYRIGHT   HOLDERS  AND
+ * CONTRIBUTORS  "AS IS"  AND  ANY  EXPRESS  OR  IMPLIED WARRANTIES,
+ * INCLUDING,  BUT  NOT  LIMITED  TO,   THE  IMPLIED  WARRANTIES  OF
+ * MERCHANTABILITY  AND  FITNESS  FOR  A   PARTICULAR   PURPOSE  ARE
+ * DISCLAIMED.
+ * IN NO EVENT SHALL THE  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+ * OF  SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR  BUSINESS  INTERRUPTION)  HOWEVER  CAUSED AND ON ANY THEORY OF
+ * LIABILITY,  WHETHER  IN  CONTRACT,   STRICT  LIABILITY,  OR  TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ * USE OF  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ */
+
+#include "uCLib.h"
+
+void sys_exit(int error_code)
+{
+	// TODO: Improve code to handle error block
+	__asm
```

Comment:
I think using assembler in this library apart from the startup/end/stack extend code is a bad idea. OSLib or _swix can be used to keep the bulk of the library in C.

Reply:
If you mean using os_swix doesn't it requires to link against risc_oslib??? where is the implementation of it? I only see the header in RISC_OSLib:os.h.

Reply:
Whoops! - OSLib it is then.

Reply:
OSLib??? woah that has a completely different way of generating the code... Look, what I am trying to build is a small and self contained CLib (the name is pretty clear). I do not want to force cascade linking to just build a util file. So not sure how this should help me. The result of this effort is simple code a util in C and link it against the uCLib and it's done, if you need more we can either add it or (obviously) you can add extra linking to other libraries on your specific util code. Makes sense?

Reply:
If you are only wanting to do CLib things and not make RISC OS specific calls in the utilities then yes. Otherwise you'll be linking in OSLib functions anyway.

Reply:
Correct, the uClib should be just good enough to build and run Utils code written in C (at least for now, because I need it mostly for the DME to build quickly some utils). For SWI calls I am planning to add a way to do it and that's it, if people wants more they probably should link against more complete libs like OSLib and/or others. 

Of course I want to reach a level where I have printf and few other utilities that would make the uCLib useful by itself, I listed the 3 categories of features in the README. But just for the sake of the conversation I want to have support for argc/argv (obviously), main function call, stack for function calls and recursion and then File I/O and minimal string library to allow uCLib to provide what is usually used to write Utilities which simply read cmd parameters, check for files and/or load them, do basic manipulation and then quit. Obviously a utility library must offer a way to access SWIs. But right now I don't think we need more than that.

Obviously if you would like to have more, you're very welcome to add it yourself or, if you prefer, open an issue with the specific tab, but I am pretty sure you'd prefer the coding approach :) - As long as we keep it (u) aka Micro incremental improvements are always welcome! :)

Reply:
I must admit I didn't get as far as reading the README, it does clear things up.